### DISCUSSION TITLE
umi4 history.push的参数为什么要改呢，我不理解

### DISCUSSION THREAD

**Discussion by allahbin-super** [umijs/umi#8252] [Q&A]
之前再umi3里面，跳转时传参数是这么写的
```
// 带参数跳转到指定路由
history.push('/list?a=b');
history.push({
  pathname: '/list',
  query: {
    a: 'b',
  },
});
```

在umi4里面就不能用query了。

```
// 带参数跳转到指定路由
history.push('/list?a=b&c=d#anchor');
history.push({
  pathname: '/list',
  search: '?a=b&c=d',
  hash: 'anchor',
});
```

这样改有什么好处呢

**Comment by Sun-Lampyridae** [comment]
我也是

**Comment by Sun-Lampyridae** [comment]
但是你可以这样传参 history.push('/library/identificationLibrary/information', record);

**Comment by xiaohuoni** [comment]
最根本的原因是 react-route 升级到 v6 之后的 api 变更，建议新的项目使用新的写法，但是在旧的项目迁移和维护时，可以使用 umi 提供的兼容方案，配置 [historyWithQuery:{}](https://umijs.org/docs/api/config#historywithquery)

**Reply by crazyair** [reply]
pathname 为什么必传了？不传不能用当前 location.pathname 吗

**Reply by xiaohuoni** [reply]
啥意思？跳转页面不写路由？

**Reply by crazyair** [reply]
以前是直接 `history.push({ search: xxx})` 会直接改当前页面的 search

**Reply by xiaohuoni** [reply]
🐮没这么用过，也没见这么用过的。

**Reply by crazyair** [reply]
反正 umi3 的时候这样用的，现在升 4，还要所有地方用 useHistry 拿一次 pathname

**Reply by fz6m** [reply]
这种情况自己封装一下然后搜索 `history.push` 改一下导入吧，可以自己封装一个自己的 `history` ，或者 hack patch 了默认的 history 方法也行：

```ts
// src/app.ts

export const modifyClientRenderOpts = (context) => {
  const h = context.history
  const originPush = h.push
  const originReplace = h.replace
  h.push = (...args) => {
    originPush.apply(h, args)
  }
  h.replace = (...args) => {
    originReplace.apply(h, args)
  }
  return context
}
```

这里面你想增加什么功能都可以。

**Reply by crazyair** [reply]
用 `modifyClientRenderOpts` 也不行，history 的 pathname 类型还是必填的，看 umi 做不做兼容，做的话我就去掉 pathname，不做的话，我就都加上 pathname 吧

**Reply by fz6m** [reply]
自行 patch history 时，里面想要接受什么参数或者执行什么内部逻辑都是可以自行定义和扩展的，所以可以解决任何 history 前后不一致的问题。

如有问题，可以给一个最小复现仓库重新开一个 issue 吧。
